_posts/2016-01-16-swarm-scales-your-containter-for-free.markdown

```diff
@@ -0,0 +1,186 @@
+---
+layout: post
+title:  "Swarm scales docker for free"
+date:   2015-12-14 10:08:27
+categories: devops
+img: /img/docker.png
+tags: devops, docker, swarm, gourmet
+summary: "Docker is an awesome tool to manage your container.
+Swarm helps you to scale your containers on more servers."
+priority: 0.6
+changefreq: yearly
+---
+
+Serverless, container, Aws lambda, [Gourmet](https://github.com/gianarb/gourmet) burs on this ideas,
+to improve my go knowledge and to work with the Docker API
+I am happy to share my idea and my tests with Swam an easy way to scale this type of application. 
+
+Gourmet exposes an HTTP api, route `/project`, method POST that wait a body similar to
+
+{% highlight json %}
+{
+    "img": "gourmet/php",
+    "source": "https://ramdom-your-source.net/gourmet.zip",
+    "env": [
+        "AWS_KEY=EXAMPLE",
+        "AWS_SECRET=",
+        "AWS_QUEUE=https://sqs.eu-west-1.amazonaws.com/test"
+    ]
+}
+{% endhighlight %}
+
+`img` is the started point docker image, `source`
+is your script, for my test I use an [easy php
+script](https://github.com/gianarb/gourmet-php-example) that send a message on SQS.
+
+Your script has a console entrypoint executables in this path `/bin/console` and
+gourmet uses it to run your program.
+
+To manage integration with docker I used `fsouza/go-dockerclient` an open source
+library written in go.
+
+{% highlight go %}
+container, err := dr.Docker.CreateContainer(docker.CreateContainerOptions{
+    "",
+    &docker.Config{
+        Image:        img,
+        Cmd:          []string{"sleep", "1000"},
+        WorkingDir:   "/tmp",
+        AttachStdout: false,
+        AttachStderr: false,
+        Env:          envVars,
+    },
+    nil,
+})
+{% endhighlight %}
+
+This is an extract of my code and I use it to create a new container.
+After it I have a container started and I use the exec docker feature to
+extract your source and to run it.
+
+{% highlight go %}
+exec, err := dr.Docker.CreateExec(docker.CreateExecOptions{
+    Container:    containerId,
+    AttachStdin:  true,
+    AttachStdout: true,
+    AttachStderr: true,
+    Tty:          false,
+    Cmd:          command,
+})
+
+if err != nil {
+    return err;
+}
+
+err = dr.Docker.StartExec(exec.ID, docker.StartExecOptions{
+    Detach:      false,
+    Tty:         false,
+    RawTerminal: true,
+    OutputStream: dr.Stream,
+    ErrorStream:  dr.Stream,
+})
+{% endhighlight %}
+
+After each build Gourmet cleans all and destroy the container.
+
+{% highlight go %}
+err := dr.Docker.KillContainer(docker.KillContainerOptions{ID: containerId})
+err = dr.Docker.RemoveContainer(docker.RemoveContainerOptions{ID: containerId, RemoveVolumes: true})
+if(err != nil) {
+    return err;
+}
+return nil
+{% endhighlight %}
+Very easy, gourmet finished here!
+
+It could be different ipotetical use cases, high separated task, run a
```

Comment:
ipotetical => hypothetical - anyway I can't understand what do you mean with this sentence